packages/components/rating/src/Rating.tsx

```diff
@@ -1 +1,180 @@
+    },
+    ref
+  ) => {
+    const inputRef = useRef<HTMLInputElement>(null)
+    const starRefList = useRef<HTMLDivElement[]>([])
+
+    const [value, setRatingValue] = useCombinedState(propValue, defaultValue, onValueChange)
+
+    const valueRef = useRef(value)
+    const isInteractive = !(disabled || readOnly)
+
+    function onStarClick(index: number) {
+      if (!inputRef.current) return
+
+      setRatingValue(index + 1)
+      valueRef.current = index + 1
+
+      inputRef.current.focus()
+      inputRef.current.setAttribute('data-clicked', '')
+    }
+
+    function onInputChange(event: ChangeEvent<HTMLInputElement>) {
+      // avoiding unnecessary calls to the onValueChange prop
+      // when the value remains unchanged
+      if (valueRef.current === Number(event.target.value)) return
+      valueRef.current = Number(event.target.value)
+
+      setRatingValue(Number(event.target.value))
+    }
+
+    function onStarMouseEnter({ currentTarget }: MouseEvent<HTMLDivElement>) {
+      const currentStarIndex = starRefList.current.findIndex(star => star === currentTarget)
+
+      const [previousStars, followingStars] = splitAt(starRefList.current, currentStarIndex + 1)
+
+      previousStars.forEach(star => star.setAttribute('data-hovered', ''))
+      followingStars.forEach(star => star.removeAttribute('data-hovered'))
+    }
+
+    const handleStarRef = useCallback((elm: HTMLDivElement | null) => {
+      if (!elm) return
+      starRefList.current.push(elm)
+    }, [])
+
+    function resetDataPartInputAttr() {
+      inputRef.current?.removeAttribute('data-clicked')
+    }
+
+    function resetDataPartStarAttr() {
+      starRefList.current.forEach(star => star.removeAttribute('data-hovered'))
+    }
+
+    return (
+      <div
+        className="relative"
+        ref={ref}
+        data-spark-component="rating"
+        {...rest}
+        onMouseLeave={resetDataPartStarAttr}
+      >
+        <input
+          name={name}
+          id={id}
+          aria-label={rest['aria-label'] ?? 'rating'}
+          ref={inputRef}
+          data-part="input"
+          className="peer absolute inset-none opacity-0"
+          type="range"
+          min="0"
+          max="5"
+          step={1}
+          disabled={disabled}
+          readOnly={readOnly}
+          value={getNearestDecimal(value)}
+          onChange={event => isInteractive && onInputChange(event)}
+          onKeyDown={resetDataPartInputAttr}
+          onBlur={resetDataPartInputAttr}
+        />
+        <div
+          className={cx(
+            size === 'lg' ? 'gap-x-md' : 'gap-x-sm',
+            'flex',
+            'peer-focus-visible:u-ring peer-[[data-part=input][data-clicked]]:shadow-none'
+          )}
+        >
+          {Array.from({ length: 5 }).map((_, index) => (
+            <RatingStar
+              disabled={disabled}
+              readOnly={readOnly}
+              size={size}
+              onClick={() => isInteractive && onStarClick(index)}
```

Comment:
This isInteractive condition can be moved into onStarClickHandler in order to be able to be memoized

Reply:
yes, but, I'm not so sure we need to memoize this function. According to the [React documentation](https://react.dev/reference/react/useCallback#should-you-add-usecallback-everywhere), you only really need to cache a function with `useCallback` in a few cases.

- Like if you're passing it as a prop to a component wrapped in `memo`, and you want to skip re-rendering if the value hasn't changed.
- Or if the function is later used as a dependency of some Hook, like if another function wrapped in `useCallback` depends on it

But in this case, it's neither of those. So, I think we can skip the memoization